src/ecs/resources/input/input_handler.rs

```diff
@@ -0,0 +1,329 @@
+                Event::ReceivedCharacter(c) => {
+                    self.text_this_frame.push(c);
+                }
+                Event::KeyboardInput(ElementState::Pressed, _, Some(key_code)) => {
+                    if self.pressed_keys.iter().all(|&k| k != key_code) {
+                        self.pressed_keys.push(key_code);
+                        self.down_keys.push(key_code);
+                    }
+                }
+                Event::KeyboardInput(ElementState::Released, _, Some(key_code)) => {
+                    let index = self.pressed_keys.iter().position(|&k| k == key_code);
+                    if let Some(i) = index {
+                        self.pressed_keys.swap_remove(i);
+                        self.released_keys.push(key_code);
+                    }
+                }
+                Event::MouseInput(ElementState::Pressed, button) => {
+                    if self.pressed_mouse_buttons.iter().all(|&b| b != button) {
+                        self.pressed_mouse_buttons.push(button);
+                        self.down_mouse_buttons.push(button);
+                    }
+                }
+                Event::MouseInput(ElementState::Released, button) => {
+                    let index = self.pressed_mouse_buttons
+                        .iter()
+                        .position(|&b| b == button);
+                    if let Some(i) = index {
+                        self.pressed_mouse_buttons.swap_remove(i);
+                        self.released_mouse_buttons.push(button);
+                    }
+                }
+                Event::MouseMoved(x, y) => {
+                    self.mouse_position = Some((x, y));
+                }
+                Event::Focused(false) => {
+                    self.pressed_keys.clear();
+                    self.pressed_mouse_buttons.clear();
+                    self.mouse_position = None;
+                }
+                _ => {}
+            }
+        }
+    }
+
+    /// Returns a string representation of all text entered this frame.
+    ///
+    /// Intended for use with text entry fields, insert this string at the cursor position
+    /// every frame.
+    pub fn text_entered(&self) -> &str {
+        self.text_this_frame.as_str()
+    }
+
+    /// Returns an iterator over all keys in the given state, does not support Released(Currently).
+    pub fn keys_that_are(&self, state: ButtonState) -> KeyCodes {
+        if let Pressed(ThisFrame) = state {
+            self.down_keys.iter()
+        }
+        else if let Pressed(Currently) = state {
+            self.pressed_keys.iter()
+        }
+        else if let Released(ThisFrame) = state {
+            self.released_keys.iter()
+        }
+        else {
+            panic!("Released(Currently) is not supported in this context.");
+        }
+    }
+
+    /// Checks if a key matches the description given by state.
+    pub fn key_is(&self, key: VirtualKeyCode, state: ButtonState) -> bool {
```

Comment:
@torkleyy Sorry, I'm not sure I understand.  A method for what?  If you're talking about a method for ButtonState then I don't quite see what the point is as it would have to take an InputHandler reference, it'd basically just be a rearrangement of this interface.  As I understand it you want

```Rust
impl ButtonState {
    pub fn key_is(&self, input: &InputHandler, key: VirtualKeyCode) -> bool {
        input.key_is(key, *self);
    }
}
```